subsys/bluetooth/host/classic/hfp_ag.c

```diff
@@ -0,0 +1,2029 @@
+	case BT_HFP_CALL_OUTGOING:
+		k_work_reschedule(&ag->deferred_work, K_SECONDS(CONFIG_BT_HFP_AG_OUTGOING_TIMEOUT));
+		break;
+	case BT_HFP_CALL_INCOMING:
+		k_work_reschedule(&ag->deferred_work, K_SECONDS(CONFIG_BT_HFP_AG_INCOMING_TIMEOUT));
+		break;
+	case BT_HFP_CALL_ALERTING:
+		k_work_reschedule(&ag->ringing_work, K_NO_WAIT);
+		k_work_reschedule(&ag->deferred_work, K_SECONDS(CONFIG_BT_HFP_AG_ALERTING_TIMEOUT));
+		break;
+	case BT_HFP_CALL_ACTIVE:
+		k_work_cancel_delayable(&ag->deferred_work);
+		break;
+	case BT_HFP_CALL_HOLD:
+		break;
+	default:
+		/* Invalid state */
+		break;
+	}
+
+	if (ag->state == BT_HFP_DISCONNECTING) {
+		int err = bt_rfcomm_dlc_disconnect(&ag->rfcomm_dlc);
+
+		if (err) {
+			LOG_ERR("Fail to disconnect DLC %p", &ag->rfcomm_dlc);
+		}
+	}
+}
+
+static struct bt_ag_tx *bt_ag_tx_alloc(void)
+{
+	/* The TX context always get freed in the system workqueue,
+	 * so if we're in the same workqueue but there are no immediate
+	 * contexts available, there's no chance we'll get one by waiting.
+	 */
+	if (k_current_get() == &k_sys_work_q.thread) {
+		return k_fifo_get(&ag_tx_free, K_NO_WAIT);
+	}
+
+	if (IS_ENABLED(CONFIG_BT_HFP_AG_LOG_LEVEL_DBG)) {
+		struct bt_ag_tx *tx = k_fifo_get(&ag_tx_free, K_NO_WAIT);
+
+		if (tx) {
+			return tx;
+		}
+
+		LOG_WRN("Unable to get an immediate free bt_ag_tx");
+	}
+
+	return k_fifo_get(&ag_tx_free, K_FOREVER);
+}
+
+static void bt_ag_tx_free(struct bt_ag_tx *tx)
+{
+	LOG_DBG("Free tx buffer %p", tx);
+
+	(void)memset(tx, 0, sizeof(*tx));
+
+	k_fifo_put(&ag_tx_free, tx);
+}
+
+static int hfp_ag_next_step(struct bt_hfp_ag *ag, bt_hfp_ag_tx_cb_t cb, void *user_data)
+{
+	struct bt_ag_tx *tx;
+
+	tx = bt_ag_tx_alloc();
+	if (tx == NULL) {
+		LOG_ERR("No tx buffers!");
+		return -ENOMEM;
+	}
+
+	tx->ag = ag;
+	tx->buf = NULL;
+	tx->cb = cb;
+	tx->user_data = user_data;
+
+	k_fifo_put(&ag_tx_notify, tx);
+
+	return 0;
+}
+
+static int hfp_ag_send(struct bt_hfp_ag *ag, struct bt_ag_tx *tx)
+{
+	return bt_rfcomm_dlc_send(&ag->rfcomm_dlc, tx->buf);
+}
+
+static int hfp_ag_send_data(struct bt_hfp_ag *ag, bt_hfp_ag_tx_cb_t cb, void *user_data,
+			    const char *format, ...)
+{
+	struct net_buf *buf;
+	struct bt_ag_tx *tx;
+	va_list vargs;
+	int err;
+
+	LOG_DBG("AG %p sending data cb %p user_data %p", ag, cb, user_data);
+
+	buf = bt_rfcomm_create_pdu(&ag_pool);
+	if (!buf) {
+		LOG_ERR("No Buffers!");
+		return -ENOMEM;
+	}
+
+	tx = bt_ag_tx_alloc();
+	if (tx == NULL) {
+		LOG_ERR("No tx buffers!");
+		net_buf_unref(buf);
+		return -ENOMEM;
+	}
+
+	LOG_DBG("buf %p tx %p", buf, tx);
+
+	tx->ag = ag;
+	tx->buf = buf;
+	tx->cb = cb;
+	tx->user_data = user_data;
+
+	va_start(vargs, format);
+	err = vsnprintk((char *)buf->data, (net_buf_tailroom(buf) - 1), format, vargs);
+	va_end(vargs);
+
+	if (err < 0) {
+		LOG_ERR("Unable to format variable arguments");
+		net_buf_unref(buf);
+		bt_ag_tx_free(tx);
+		return err;
+	}
+
+	net_buf_add(buf, err);
+
+	k_sem_take(&ag->locker, K_FOREVER);
```

Comment:
As I suggested earlier, I think it'd be better to have a more natural locking wrapper for this, since semaphore taking/giving is not an intuitive locking operation. So something like:
```c
static void hfp_ag_lock(struct bt_hfp_ag *ag)
{
	k_sem_take(&ag->lock, K_FOREVER);
}

static void hfp_ag_unlock(struct bt_hfp_ag *ag)
{
	k_sem_give(&ag->lock);
}
```

Since you now have a lock for protecting the `bt_hf_ag` struct, you should use it everywhere where there is a chance that some other thread may be accessing it at the same time.

Reply:
udpated